Build with vite (and rollup)

kilobyte2007 opened the pull request:
I tried to experiment a little with building with vite so we can get rid of webpack. There was a bit of complication with the lib build as vite does not allow multiple entries but I have found an elegant way of doing it.

Please take a look once you have some time and let me know what you think.

kilobyte2007 committed:
Implemented the building with vite (rollup) and got rid of unnecessary deps

kilobyte2007 committed:
Implemented the building with vite (rollup) and got rid of unnecessary deps (typo fix)

kilobyte2007 committed:
Syntax fixes

kilobyte2007 committed:
Implemented the building with vite (rollup) and got rid of unnecessary deps

JosephusPaye commented:
Thanks for this! I'll have a closer look soon.

kilobyte2007 commented:
Hey, @JosephusPaye any updates on this and the other PR and release of the new version?

Don't want to be rude or ungrateful but are there any plans to maintain and move the project forward?
If the project will remain dormant maybe there is a sense for us to fork it and maintain a fork where we can implement new stuff and improve it.

JosephusPaye commented:
Hi @kilobyte2007,

I'd definitely like the project to move forward as I still use it myself in production applications. I'd love to review the PRs you submitted and have been unsuccessfully trying to find time to do so in the last few weeks. I'm happy to give you commit access to the repo, and make you a maintainer, so I'm no longer a blocker. Let me know if that's something you're interested in.

Thanks!

JosephusPaye reviewed (changes requested):
I just cloned the PR and tested it. These are the things stood out to me:

- It looks like Vite is bundling Vue into each built file, including the `dist/` files and `lib/` files. This increases the size of each file considerably, and is the biggest issue I can see. Here you can see the difference in the file sizes (left is before with webpack, right is after with vite).

  ![image](https://user-images.githubusercontent.com/5924865/207813311-e449bb49-5e40-4a6e-a14c-f5ca109f5905.png)

  _See the size for UiTab for example, it went from 8KB to 188KB._

- We're no longer building both minified and non-minified versions of the dist and lib files. It appears `vite build` always minifies files, [though this can be configured](https://vitejs.dev/config/build-options.html#build-minify). I'd like to keep the non-min builds because they're useful for debugging when you're not importing Keen UI from the `src/` folder.
- Built lib files no longer have the banner
- This might just be me, but sometimes running `yarn build:lib` doesn't build all the 34 components. Re-running usually fixes it. You can try it with `rm -rf lib/ && yarn build:lib && ls lib/ | wc -l` to see if the same thing happens for you. If it works correctly the output should be 34.

kilobyte2007 commented:
Hey @JosephusPaye,
Thanks for reviewing this, I did not think to take a look at file sizes. Good catch. I'll take a look!

kilobyte2007 committed:
Building with vite - bugfixes (minification, bundle size, banner, missing components)

kilobyte2007 commented:
Hey,
I've updated the PR, all 4 issues are fixed now:
1. I have configured rollup to exclude `vue` from the build and use the global `Vue`
2. Now both minified and non-minified versions are being built when in production mode
3. Fixed the banner, turns out we have to specify the `outDir` to the plugin as well
4. There was indeed an issue as vite/rollup was cleaning the directory at random times

I've run some tests and it seems to work fine, but please let me know if you find any issues!

P.S. On making me a maintainer - I would be happy to join forces. Let me know what's going to be needed from me in this case and also what the workflow is going to look like from your perspective.

JosephusPaye commented on build/build.lib.js:
```diff
@@ -39,5 +39,9 @@ const entries = [
 
 entries.forEach(async entry => {
-    await build(profileLibConfig({ entry, mode: process.env.NODE_ENV || 'development' }));
+    if (process.env.NODE_ENV === 'production') {
+        await build(getLibConfig({ entry, minify: true }));
+    }
+
+    await build(getLibConfig({ entry, minify: false }));
```
```suggestion
    await build(getLibConfig({ entry, minify: process.env.NODE_ENV === 'production' }));
```

kilobyte2007 replied:
This won't work, because in order to both build the minified and the non-minified versions, we need to run the script twice cause vite/rollup only builds one version at the times, which is different from how webpack works.

kilobyte2007 replied:
I think a solution would be to change the commands so it builds in both prod and non-prod mode:
```
    "prod": "cross-env NODE_ENV=development npm run build:all && cross-env NODE_ENV=production npm run build:all",
```

What do you think? This way it would build both versions of the bundle.

JosephusPaye replied:
Yeah makes sense. I am planning on improving the build scripts after this PR, so we can leave that off till then.

This is what I have on a local branch (for webpack, which will change to vite after this PR):

<details>
<summary>View image</summary>

![image](https://user-images.githubusercontent.com/5924865/209622699-9f5c86c6-c802-4450-a6ad-5f0352637309.png)

</details>

kilobyte2007 replied:
Ok, so I tried to implement something similar like this. Please take a look. Because vite works a bit differently I've had to organize the commands a bit differently, but now at least there's consistency.

Also:
* Fixed some styling issues
* Added autoprefixer
* Excluded vue from being bundled

Please let me know what you think!

JosephusPaye commented on build/vite.config.lib.provider.js:
```diff
@@ -2,27 +2,38 @@ const vue = require('@vitejs/plugin-vue');
+
     return {
-        plugins: [
+        plugins: [,
```
```suggestion
        plugins: [
```

JosephusPaye reviewed (changes requested):
Thanks for the changes! Here is another review.

- `lib/` files now build both min and non-min versions, and it looks like most of the file sizes are now smaller 🎉

  <details><summary>Size comparison</summary>

  ```
  Filename                          Size before   Size after        Delta
  -----------------------------------------------------------------------
  UiAlert.js                           50.41 kb     20.28 kb    -30.13 kb
  UiAlert.min.js                       18.99 kb     12.38 kb     -6.60 kb
  UiAutocomplete.js                    48.94 kb     23.40 kb    -25.54 kb
  UiAutocomplete.min.js                19.58 kb     14.37 kb     -5.21 kb
  UiButton.js                         240.94 kb    145.78 kb    -95.16 kb
  UiButton.min.js                      88.78 kb     73.15 kb    -15.63 kb
  UiCalendar.js                       282.04 kb    166.81 kb   -115.23 kb
  UiCalendar.min.js                   102.86 kb     85.98 kb    -16.87 kb
  UiCheckbox.js                        25.04 kb      8.21 kb    -16.83 kb
  UiCheckbox.min.js                    10.40 kb      5.76 kb     -4.64 kb
  UiCheckboxGroup.js                   38.01 kb     16.46 kb    -21.55 kb
  UiCheckboxGroup.min.js               15.84 kb     10.90 kb     -4.94 kb
  UiCloseButton.js                     40.17 kb     12.95 kb    -27.22 kb
  UiCloseButton.min.js                 14.19 kb      7.79 kb     -6.40 kb
  UiCollapsible.js                     43.02 kb     15.96 kb    -27.05 kb
  UiCollapsible.min.js                 15.15 kb      9.49 kb     -5.66 kb
  UiConfirm.js                        276.64 kb    166.22 kb   -110.43 kb
  UiConfirm.min.js                    103.30 kb     91.36 kb    -11.94 kb
  UiDatepicker.js                     357.82 kb    223.56 kb   -134.26 kb
  UiDatepicker.min.js                 144.40 kb    129.71 kb    -14.68 kb
  UiDatepickerCalendar.js             291.86 kb    174.93 kb   -116.93 kb
  UiDatepickerCalendar.min.js         109.01 kb     97.07 kb    -11.94 kb
  UiFab.js                            190.98 kb    111.60 kb    -79.38 kb
  UiFab.min.js                         61.59 kb     54.62 kb     -6.98 kb
  UiFileupload.js                      46.27 kb     19.81 kb    -26.46 kb
  UiFileupload.min.js                  18.69 kb     12.58 kb     -6.11 kb
  UiIcon.js                            19.26 kb      2.41 kb    -16.84 kb
  UiIcon.min.js                         5.74 kb      1.49 kb     -4.26 kb
  UiIconButton.js                     241.02 kb    145.47 kb    -95.55 kb
  UiIconButton.min.js                  88.99 kb     73.34 kb    -15.65 kb
  UiMenu.js                            55.08 kb     20.05 kb    -35.03 kb
  UiMenu.min.js                        17.93 kb     11.63 kb     -6.30 kb
  UiModal.js                           62.96 kb     27.06 kb    -35.90 kb
  UiModal.min.js                       23.74 kb     16.64 kb     -7.11 kb
  UiPopover.js                        182.35 kb    110.06 kb    -72.29 kb
  UiPopover.min.js                     58.95 kb     53.27 kb     -5.68 kb
  UiPreloader.js                       21.48 kb      3.29 kb    -18.20 kb
  UiPreloader.min.js                    7.92 kb      2.26 kb     -5.66 kb
  UiProgressCircular.js                24.33 kb      6.99 kb    -17.33 kb
  UiProgressCircular.min.js             9.43 kb      4.58 kb     -4.85 kb
  UiProgressLinear.js                  23.04 kb      4.83 kb    -18.21 kb
  UiProgressLinear.min.js               9.04 kb      3.34 kb     -5.70 kb
  UiRadio.js                           24.29 kb      7.45 kb    -16.85 kb
  UiRadio.min.js                        9.78 kb      5.02 kb     -4.75 kb
  UiRadioGroup.js                      35.45 kb     15.03 kb    -20.42 kb
  UiRadioGroup.min.js                  14.57 kb      9.80 kb     -4.76 kb
  UiRippleInk.js                       25.82 kb      6.95 kb    -18.87 kb
  UiRippleInk.min.js                    9.00 kb      3.87 kb     -5.13 kb
  UiSelect.js                         242.45 kb    150.86 kb    -91.59 kb
  UiSelect.min.js                      86.79 kb     78.00 kb     -8.80 kb
  UiSlider.js                          40.57 kb     18.21 kb    -22.36 kb
  UiSlider.min.js                      16.72 kb     11.33 kb     -5.40 kb
  UiSnackbar.js                       248.88 kb    149.60 kb    -99.28 kb
  UiSnackbar.min.js                    91.83 kb     75.65 kb    -16.18 kb
  UiSnackbarContainer.js              258.00 kb    154.15 kb   -103.85 kb
  UiSnackbarContainer.min.js           94.77 kb     78.19 kb    -16.58 kb
  UiSwitch.js                          26.01 kb      7.75 kb    -18.27 kb
  UiSwitch.min.js                      10.42 kb      5.18 kb     -5.25 kb
  UiTab.js                             21.06 kb      3.45 kb    -17.62 kb
  UiTab.min.js                          6.33 kb      2.12 kb     -4.21 kb
  UiTabs.js                            53.45 kb     21.51 kb    -31.94 kb
  UiTabs.min.js                        19.75 kb     12.96 kb     -6.79 kb
  UiTextbox.js                         46.14 kb     25.17 kb    -20.97 kb
  UiTextbox.min.js                     19.35 kb     14.76 kb     -4.60 kb
  UiToolbar.js                        259.62 kb    154.98 kb   -104.64 kb
  UiToolbar.min.js                     97.46 kb     79.59 kb    -17.87 kb
  UiTooltip.js                        165.62 kb    100.08 kb    -65.55 kb
  UiTooltip.min.js                     52.59 kb     47.75 kb     -4.83 kb
  ```

- Seems like we're still not building the non-min `dist/` files, only `keen-ui.min.css` and `keen-ui.min.js`. Also, I think Vue is still bundled in the built `keen-ui.min.js` file, as it it's now much larger:

  <details><summary>Size comparison</summary>

  ```
  Filename                          Size before   Size after        Delta
  -----------------------------------------------------------------------
  keen-ui.css                         111.03 kb       (none)       (none)
  keen-ui.js                          513.29 kb       (none)       (none)
  keen-ui.min.css                     107.07 kb     87.35 kb    -19.71 kb
  keen-ui.min.js                      176.51 kb    346.66 kb    170.15 kb (larger)
  ```
  </details>

- With Vite I don't think we need the `.babelrc` file in the root. We can remove it now. The default browsers supported by Vite are good for Keen UI. Note for self: need to document this in the changelog, and provide some info for those we want to support older browsers.

- I think we'll need to add `autoprefixer` to the `postcss` options in the Vite config.

---

About becoming maintainer: you'll have commit access to merge your PRs, and eventually I'll add you to npm to publish releases. In the beginning I'd like to still have a quick look at the PRs, and if you want to make bigger changes, I'd like for us to quickly discuss them first.

kilobyte2007 committed:
Update build/vite.config.lib.provider.js

Fix unnecessary comma after copy-pasting

Co-authored-by: Josephus Paye II <j.paye96@gmail.com>

kilobyte2007 committed:
Building with vite - improve commands and separate by mode

kilobyte2007 committed:
Building with vite - unbundle vue from dist

JosephusPaye reviewed (changes requested):
Looks great! Almost every output file is now smaller, in some cases by more than 100kb!

Only one thing left and we can merge this: add the meta tags to the Vite html entry point for the docs, from here: https://github.com/JosephusPaye/Keen-UI/blob/9d6ff42b02569a2fc5541faa275339b06eae1cfa/docs/index.html#L12-L21

Many thanks for your continued work on this.

kilobyte2007 committed:
Building with vite - add meta tags and descriptions to the docs entry file

kilobyte2007 commented:
Oh, totally forgot to migrate the HTML file from the previous version. I've brought it to the previous format now, please take a look!

JosephusPaye reviewed (approved):
Looks Good To Me!
